Blog post of the powershell lab (#1552)
oktadeveloper/okta-blog: _source/_posts/2025-05-07-terraform-powershell-lab.md

---
layout: blog_post
title: "Scaling Okta App Management by Importing Data from PowerShell into Terraform"
author: edunham
by: advocate
communities: [devops,.net]
description: "Hands-on guide to using Okta Terraform and Powershell."
tags: [terraform, powershell]
tweets:
- ""
- ""
- ""
image: blog/terraform-powershell-lab/image.jpg
type: awareness
---

At Oktane 2024, we offered a lab session to teach the basics of automating Okta administration tasks with Terraform and PowerShell. This post walks you through the lab content so you can follow along at your own pace to get some hands-on experience managing an Okta tenant with both tools. 

> This lab includes some Scenario sections to give realistic examples of a situation where you might perform each task. You can ignore these sections if you don't find them helpful. 

Reviewer: ```suggestion
Each task in this lab includes realistic scenarios of situations where you may want to consider applying these concepts. You can ignore these sections if you don't find them helpful. 
```

> ## Background

Reviewer: ```suggestion
## Use PowerShell and Terraform to configure Okta
```

Thinking of ways to add PowerShell, Terraform, and Okta in a header to boost SEO. Please apply this idea to other headers appropriately

> PowerShell is an excellent introductory tool for general automation tasks. It is easier to set up but can require extensive customization to integrate with a more complete infrastructure-as-code solution. In this lab, the focus will be on reading data with PowerShell, but it is a general-purpose tool that can read from and write to Okta when configured with the appropriate scopes.

Reviewer: Should we link out to PowerShell and Terraform here?

Terraform is a specialized tool for developing infrastructure as code. Terraform requires more setup steps than PowerShell, but adding Okta management to an existing Terraform deployment is more straightforward than designing a new solution with PowerShell.

## How to create your lab environment

You can follow these steps on any computer where you're able to install the following dependencies: 

- Install [PowerShell](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-7.4). This lab was tested on PowerShell 7.4.

> - Install the [official Okta PowerShell module](https://github.com/okta/okta-powershell-cli) from your preferred PowerShell module location. This lab weas tested on v1.0.3. 

Reviewer: ```suggestion
- Install the [official Okta PowerShell module](https://github.com/okta/okta-powershell-cli) from your preferred PowerShell module location. This lab was tested on v1.0.3. 
```

- Install [Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli). This lab was tested on v1.11.3. 
- Create an Okta org by [signing up](https://developer.okta.com/signup/) for Workforce Identity Cloud.

> **Note:** Free trial accounts expire after 30 days, but developer accounts do not expire. 

Reviewer: ```suggestion
**Note:** Free trial accounts expire after 30 days, but Okta Developer Edition accounts do not expire. 
```

To use Terraform, you'll run commands in a terminal. You can even use a PowerShell window if you don't have a favorite terminal! 

> If this is your first time writing code, you might want to install an [IDE](https://en.wikipedia.org/wiki/Integrated_development_environment) like [Visual Studio](https://code.visualstudio.com/docs/introvideos/basics) so that you can navigate the files you're writing, detect errors using plugins, and execute commands using the built-in terminal. 

Reviewer: ```suggestion
If this is your first time writing code, you might want to install an [IDE](https://en.wikipedia.org/wiki/Integrated_development_environment) like [Visual Studio Code](https://code.visualstudio.com/docs/introvideos/basics) so that you can navigate the files you're writing, detect errors using plugins, and execute commands using the built-in terminal. 
```

Visual Studio is a different IDE

### Accessing Terraform

In a production setting, you would collaborate on your Terraform files with colleagues using version control. If anyone at your organization uses Terraform, you should set it up according to their advice on best practices. 

Note that Terraform creates a `.tfstate` file. Each real-world resource should only be represented in one `.tfstate` file. If you're collaborating on a Terraform codebase, products like [Terraform Cloud](https://www.hashicorp.com/resources/what-is-terraform-cloud) are available to help you manage shared access to `.tfstate` files. 

You'll save a private key alongside your code to simplify the lab. In production, you should store keys in your organization's secrets management solution. Because private keys are secrets, follow your organization's best practices for secrets management when dealing with them.

## Set up your Terraform Application

> 1. Go to Applications > Applications in the Okta admin console.

Reviewer: ```suggestion
1. Go to **Applications** > **Applications** in the Okta Admin Console.
```

Style guide calls for bold UI text and reference to the admin console is "Okta Admin Console"

2. Create an app integration for the Terraform lab.

>     a. Select Create App Integration.

Reviewer: Please bold UI text for all remaining steps here

5. Complete the configuration steps.
    
    a. From the Okta Admin console, select Done.
    
    b. Select Save.
    
    c. In the "Existing client secrets will no longer be used" popup, select Save.

6. From the Okta API Scopes tab, grant the following scopes:
    - `okta.appGrants.manage`
    - `okta.apps.manage`
    - `okta.oauthIntegrations.manage`
    - `okta.policies.read`
    - `okta.users.manage`

7. Modify the admin roles for the Terraform lab app.
    
    a. Go to the Admin Roles tab.
    
    b. Select Edit Assignments.
    
    c. Select the role of Super Adminstrator from the dropdown menu.
    
    d. Select Save Changes.

8. In your VM, create the file `main.tf`:

   ```hcl
   terraform {
     required_providers {
       okta = {
         source = "okta/okta"
       }
     }
   }

   variable "org_id" {
     default = "ORGID"
   }

   provider "okta" {
     org_name  = var.org_id
     base_url  = "oktapreview.com"
     client_id = "CLIENTID"
     scopes = ["okta.apps.manage", "okta.appGrants.manage",
       "okta.oauthIntegrations.manage", "okta.users.manage",
     "okta.policies.read"]
     private_key = file("key.pem")
   }

> 

Reviewer: ```suggestion
```

   ```
   
9. In your `main.tf` file, replace `ORGID` with the subdomain for your Okta org.

   **Note:** The subdomain is between `https://` and `.oktapreview.com`. In the example below, the subdomain is `oktaice0000000`.

10. In your `main.tf` file, replace `CLIENTID` with the client ID of your Terraform lab app.

    **Note:** Go to Applications > Applications to find the client ID of your Terraform lab app.

11. Save the `main.tf` file.
12. Initiate and run the terraform plan.
    
    a. In your terminal, run the following commands:
      
      ```
      terraform init
      terraform plan
      ```

>     b. Wait for the terraform plan command to return "No changes. Your infrastructure matches the configuration."

Reviewer: Did you verify rendering here? GitHub is showing that something doesn't match. It might be just a GitHub thing, but I'm guessing you need another newline after the codeblock??

Reviewer: Should "terraform" by capitalized?

## Set up and use Okta's PowerShell Module

In this lab, we'll use Terraform to create an application that PowerShell can use to read Okta data safely. If you want to use Okta's PowerShell module without managing it in Terraform, follow [the usage guide](https://github.com/okta/okta-powershell-cli?tab=readme-ov-file#usage-guide) instead. 

> **Enterprise scenario 2**
> 

> > An Okta Ice intern uses PowerShell to generate a newsletter with statistics about the company. They have asked you to help them access data about how the company uses Okta.  You want to give their script read access to various data from Okta by setting up an application for PowerShell and testing it out. Since they are already using PowerShell, the Okta PowerShell Module is a perfect choice for the task!

Reviewer: ```suggestion
> An Okta Ice intern uses PowerShell to generate a newsletter with statistics about the company. They asked you to help them access data about how the company uses Okta.  You want to give their script read access to various data from Okta by setting up an application for PowerShell and testing it out. Since they are already using PowerShell, the Okta PowerShell Module is a perfect choice for the task!
```

passivity

### Accessing PowerShell

> You can access PowerShell 7 by launching it from the shortcut generated during its installation or with the Visual Studio Code built-in terminal. 

Reviewer: Is this PowerShell version 7? or is the 7 a mistype?

3. Type `yes` when prompted.
4. Wait for the `terraform apply` to complete.
5. From the `Creation complete after` output, copy the value of the `okta_app_oauth` id.

   **Note:** PowerShell and Terraform now have separate applications in your Okta organization. The PowerShell application's ID can also be found in the Okta admin console under Applications -> Applications. 

6. In PowerShell, run these commands:

   **Note:** Replace the {yourOktaDomain} with the entire domain for your Okta org. For example, `oktaice0000000.oktapreview.com`. Replace the ID with the value you copied above.

   ```powershell
   $Configuration = Get-OktaConfiguration
   $Configuration.BaseUrl = "https://{yourOktaDomain}"
   $Configuration.ClientId = "id"
   $Configuration.Scope = "okta.apps.read okta.domains.read okta.groups.read okta.logs.read okta.oauthIntegrations.read okta.orgs.read okta.userTypes.read okta.users.read"
   Invoke-OktaEstablishAccessToken
   ```
 
 7. The `Invoke-OktaEstablishAccessToken` command displays a URL. Open the link in a web browser.
 8. Authenticate to your Okta Training Org when prompted.
 9. In your PowerShell session, run the command `Invoke-OktaListApplications`

> ## Extra credit

Reviewer: Any way to reinforce keywords?

> You have set up the PowerShell application with various scopes. What interesting information about your organization can you retrieve using these scopes? Will these scopes allow you to make any changes to the Okta organization? Try some PowerShell commands to create or destroy resources and see what happens!

Reviewer: ```suggestion
You've set up the PowerShell application with various scopes. What interesting information about your organization can you retrieve using these scopes? Will these scopes allow you to make any changes to the Okta organization? Try some PowerShell commands to create or destroy resources and see what happens!
```

styleguide prefer contractions

## Manage Terraform using Terraform

Now, we'll use data gathered by Powershell to improve your Terraform configuration.

> If you aren't using PowerShell, you can get the Terraform integration ID from the Okta admin console instead. 

Reviewer: ```suggestion
If you aren't using PowerShell, you can get the Terraform integration ID from the Okta Admin Console instead. 
```

1. 	In your PowerShell session from the previous section, run the command `powershell Invoke-OktaListApplications`

2. 	In the resulting list, find the ID of your Terraform app. Note that the name you chose, such as "Terraform lab," will be shown in the label field. 

3. 	Add the following code to your `main.tf`, substituting the application ID you got from PowerShell: 

    ``` hcl
    import {
      to = okta_app_oauth.tf
      id="FIXME"
    }
    ```

4. 	Save `main.tf`

5.	In your terminal, run the command `terraform plan --generate-config-out tf-app-config.tf` 

6.	Remove the `import{...}` block from `main.tf` and save `main.tf`

7.	In your terminal, run the command: `terraform apply` 

8.	Open the file `tf-app-config.tf` 

9.	Remove the comments from the top of the file (single-line comments in Terraform start with `#` or `//`)

10.	Remove the lines from the resource block that do not represent properties you plan to manage. You can safely remove any null field. 

11. Copy the remaining contents of `tf-app-config.tf` into `main.tf`

12.	Remove the file `tf-app-config.tf`

13.	To complete the import, run `terraform apply`

### Extra credit 1

> Consult the Okta Terraform Provider documentation at https://registry.terraform.io/providers/okta/okta/latest/docs. 

Reviewer: Should this be a link?

Find an interesting resource and use Terraform to create that resource in your Okta training organization. 
After creating a resource in your Okta organization, you can view it in the admin console. 
 
For more examples of things you can do in Terraform, see the [Terraform Enterprise Maturity Workshop](/blog/2023/07/28/terraform-workshop). 

### Extra credit 2

> Now that you are managing Terraform's Okta application in Terraform, you can modify it with your code. Add the scope `okta.groups.manage` to your Terraform provider. 

Reviewer: ```suggestion
Now that you're managing Terraform's Okta application in Terraform, you can modify it with your code. Add the scope `okta.groups.manage` to your Terraform provider. 
```

  app_id = okta_app_oauth.ps.id
  issuer = "https://${var.org_id}.oktapreview.com"
  scopes = ["okta.apps.read", "okta.domains.read", "okta.groups.read", "okta.logs.read", "okta.oauthIntegrations.read", "okta.orgs.read", "okta.userTypes.read", "okta.users.read"]
}

resource "okta_app_oauth" "tf" {

  app_links_json = jsonencode({
    oidc_client_link = true
  })
  app_settings_json              = jsonencode({})
  authentication_policy          = "abc123"
  client_id                      = "abc123"
  consent_method                 = "REQUIRED"
  grant_types                    = ["client_credentials"]
  implicit_assignment            = false
  issuer_mode                    = "DYNAMIC"
   label                          = "Terraform lab"
  login_mode                     = "DISABLED"
   pkce_required                  = false
  policy_uri                     = null
    response_types                 = ["token"]
  token_endpoint_auth_method     = "private_key_jwt"
  type                           = "service"
  user_name_template             = "$${source.login}"
    user_name_template_type        = "BUILT_IN"
  wildcard_redirect              = "DISABLED"
  jwks {
   ...
  }
}
```

> ## Learn more about PowerShell

Reviewer: 👍🏼
